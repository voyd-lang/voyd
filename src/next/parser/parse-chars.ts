import { Expr, Form } from "./ast/index.js";
import { CharStream } from "./char-stream.js";
import { Lexer } from "./lexer.js";
import { getReaderMacroForToken } from "./reader-macros/index.js";
import { Token } from "./token.js";

export type ParseCharsOpts = {
  nested?: boolean;
  terminator?: string;
};

export const parseChars = (
  file: CharStream,
  opts: ParseCharsOpts = {}
): Form => {
  const lexer = new Lexer();
  const location = file.currentSourceLocation();
  const elements: Expr[] = [];

  while (file.hasCharacters) {
    const token = lexer.tokenize(file);

    if (token.value === opts.terminator) {
      break;
    }

    const result = processWithReaderMacro(token, file, elements.at(-1));
    if (result) elements.push(result);
  }

  location.setEndToStartOf(file.currentSourceLocation());
  const form = new Form({ location, elements });
  return opts.nested ? form : new Form(["ast", form]);
};

/** Returns true if token was matched with and processed by a macro  */
const processWithReaderMacro = (
  token: Token,
  file: CharStream,
  last?: Expr
) => {
  const readerMacro = getReaderMacroForToken(token, last, file.next);
  if (!readerMacro) return undefined;

  return readerMacro(file, {
    token,
    reader: (file, terminator) =>
      parseChars(file, {
        nested: true,
        terminator,
      }),
  });
};
