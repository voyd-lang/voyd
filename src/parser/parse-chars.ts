import { Expr } from "../syntax-objects/expr.js";
import { List } from "../syntax-objects/list.js";
import { CharStream } from "./char-stream.js";
import { Lexer } from "./lexer.js";
import { getReaderMacroForToken } from "./reader-macros/index.js";
import { Token } from "./token.js";

export type ParseCharsOpts = {
  nested?: boolean;
  terminator?: string;
  parent?: Expr;
  lexer?: Lexer;
};

export const parseChars = (
  file: CharStream,
  opts: ParseCharsOpts = {}
): List => {
  const lexer = opts.lexer ?? new Lexer();
  const list = new List({
    location: file.currentSourceLocation(),
    parent: opts.parent,
  });

  while (file.hasCharacters) {
    const token = lexer.tokenize(file);

    if (token.value === opts.terminator) {
      break;
    }

    if (processWithReaderMacro(token, list.last(), file, list, lexer)) {
      continue;
    }
  }

  list?.setEndLocationToStartOf(file.currentSourceLocation());
  return opts.nested ? list : new List(["ast", list]);
};

/** Returns true if token was matched with and processed by a macro  */
const processWithReaderMacro = (
  token: Token,
  prev: Expr | undefined,
  file: CharStream,
  list: List,
  lexer: Lexer,
) => {
  const readerMacro = getReaderMacroForToken(token, prev, file.next);
  if (!readerMacro) return undefined;

  const result = readerMacro(file, {
    token,
    reader: (file, terminator) =>
      parseChars(file, {
        nested: true,
        terminator,
        lexer,
      }),
  });

  if (result) list.push(result);
  return true;
};
