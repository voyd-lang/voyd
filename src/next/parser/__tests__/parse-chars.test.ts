import { parseChars } from "../parse-chars.js";
import { voydFile } from "./fixtures/parse-text-voyd-file.js";
import { test } from "vitest";

test("should parse the example file into a raw ast", async ({ expect }) => {
  const parserOutput = parseChars(voydFile);
  expect(parserOutput).toMatchSnapshot();
});
