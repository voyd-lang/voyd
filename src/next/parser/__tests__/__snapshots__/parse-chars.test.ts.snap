// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should parse the example file into a raw ast 1`] = `
[
  "ast",
  [
    "
",
    "fn",
    " ",
    "fib",
    [
      "paren",
      [
        "n",
        ":",
        " ",
        "i32",
      ],
    ],
    " ",
    "->",
    " ",
    "i32",
    "
",
    "  ",
    "if",
    " ",
    "n",
    " ",
    "<=",
    " ",
    "1",
    " ",
    "then",
    ":",
    "
",
    "  ",
    "  ",
    "n",
    "
",
    "  ",
    "else",
    ":",
    "
",
    "  ",
    "  ",
    "fib",
    [
      "paren",
      [
        "n",
        " ",
        "-",
        " ",
        "1",
      ],
    ],
    " ",
    "+",
    " ",
    "fib",
    [
      "paren",
      [
        "n",
        " ",
        "-",
        " ",
        "2",
      ],
    ],
    "
",
    "
",
    "fn",
    " ",
    "main",
    [
      "paren",
      [],
    ],
    "
",
    "  ",
    "let",
    " ",
    "x",
    " ",
    "=",
    " ",
    "10",
    " ",
    "+",
    "
",
    "  ",
    "  ",
    "20",
    " ",
    "+",
    "
",
    "  ",
    "  ",
    "30",
    "
",
    "
",
    "  ",
    "let",
    " ",
    "y",
    " ",
    "=",
    " ",
    "if",
    " ",
    "x",
    " ",
    ">",
    " ",
    "10",
    "
",
    "  ",
    "  ",
    "then",
    ":",
    "
",
    "  ",
    "  ",
    "  ",
    "10",
    "
",
    "  ",
    "  ",
    "else",
    ":",
    "
",
    "  ",
    "  ",
    "  ",
    "20",
    "
",
    "
",
    "  ",
    "call",
    " ",
    "this",
    " ",
    "while",
    " ",
    [
      "paren",
      [],
    ],
    " ",
    "=>",
    " ",
    "if",
    " ",
    "x",
    " ",
    ">",
    " ",
    "10",
    " ",
    "then",
    ":",
    "
",
    "  ",
    "  ",
    "x",
    " ",
    "-=",
    " ",
    "1",
    "
",
    "  ",
    "else",
    ":",
    "
",
    "  ",
    "  ",
    "x",
    " ",
    "+=",
    " ",
    "1",
    "
",
    "
",
    "  ",
    "let",
    " ",
    "n",
    " ",
    "=",
    " ",
    "if",
    " ",
    "args",
    ".",
    "len",
    [
      "paren",
      [],
    ],
    " ",
    ">",
    " ",
    "1",
    " ",
    "then",
    ":",
    "
",
    "  ",
    "  ",
    "console",
    ".",
    "log",
    [
      "paren",
      [
        [
          "new_string",
          [
            [
              "object_literal",
              [
                [
                  "label",
                  [
                    "from",
                    [
                      "FixedArray",
                      [
                        [
                          "generics",
                          [
                            "i32",
                          ],
                        ],
                        "72",
                        "101",
                        "121",
                        "32",
                        "116",
                        "104",
                        "101",
                        "114",
                        "101",
                        "33",
                      ],
                    ],
                  ],
                ],
              ],
            ],
          ],
        ],
      ],
    ],
    "
",
    "  ",
    "  ",
    "args",
    ".",
    "at",
    [
      "paren",
      [
        "1",
      ],
    ],
    ".",
    "parseInt",
    [
      "paren",
      [],
    ],
    ".",
    "unwrap",
    [
      "paren",
      [],
    ],
    "
",
    "  ",
    "else",
    ":",
    "
",
    "  ",
    "  ",
    "10",
    "
",
    "
",
    "  ",
    "let",
    " ",
    "x2",
    " ",
    "=",
    " ",
    "10",
    "
",
    "  ",
    "let",
    " ",
    "z",
    " ",
    "=",
    " ",
    "nothing",
    [
      "paren",
      [],
    ],
    "
",
    "  ",
    "let",
    " ",
    "test_spacing",
    " ",
    "=",
    " ",
    "fib",
    " ",
    "n",
    "
",
    "  ",
    "let",
    " ",
    "result",
    " ",
    "=",
    " ",
    "fib",
    [
      "paren",
      [
        "n",
      ],
    ],
    "
",
  ],
]
`;
